File: patch_pinterest.go
package main

import (
	"fmt"
	"io/ioutil"
	"strings"
)

func main() {
	// patch internal/commands/commands.go
	content, _ := ioutil.ReadFile("internal/commands/commands.go")
	str := string(content)
	str = strings.Replace(str, "results = pinterest.SearchPinterest(last.Query, last.Aspect)", "results = pinterest.SearchPinterest(last.Query, last.Aspect, last.Count)", 1)
	ioutil.WriteFile("internal/commands/commands.go", []byte(str), 0644)

	// patch main.go
	content, _ = ioutil.ReadFile("main.go")
	str = string(content)
	str = strings.Replace(str, "results = pinterest.SearchPinterest(req.Query+\" gif\", \"gif\")", "results = pinterest.SearchPinterest(req.Query+\" gif\", \"gif\", overrideCount)", 1)
	str = strings.Replace(str, "results = pinterest.SearchPinterest(req.Query+suffix, aspect)", "results = pinterest.SearchPinterest(req.Query+suffix, aspect, overrideCount)", 1)
	str = strings.Replace(str, "results = pinterest.SearchPinterestMedia(req.Query, \".mp4\")", "results = pinterest.SearchPinterestMedia(req.Query, \".mp4\", overrideCount)", 1)
	ioutil.WriteFile("main.go", []byte(str), 0644)

	// patch internal/pinterest/media_search.go
	content, _ = ioutil.ReadFile("internal/pinterest/media_search.go")
	str = string(content)
	str = strings.Replace(str, "func SearchPinterestMedia(query string, ext string) []PinResult {", "func SearchPinterestMedia(query string, ext string, count int) []PinResult {", 1)
	str = strings.Replace(str, "pins := SearchPinterest(searchQuery, \"all\")", "pins := SearchPinterest(searchQuery, \"all\", count)", 1)
	ioutil.WriteFile("internal/pinterest/media_search.go", []byte(str), 0644)

	// patch internal/pinterest/pinterest.go
	content, _ = ioutil.ReadFile("internal/pinterest/pinterest.go")
	str = string(content)
	str = strings.Replace(str, "pins := SearchPinterest(\"matching icons \"+query, \"all\")", "pins := SearchPinterest(\"matching icons \"+query, \"all\", 10)", 1)
	ioutil.WriteFile("internal/pinterest/pinterest.go", []byte(str), 0644)

	fmt.Println("Done")
}

File: patch_pinterest.py
import re

with open("internal/pinterest/pinterest.go", "r") as f:
    content = f.read()

# 1. Update SearchState and SetLastSearch
target_state = """type SearchState struct {
	Query       string
	Aspect      string
	Count       int
	IsVisual    bool
	Base64Image string
}

func SetLastSearch(chatID, query, aspect string, count int, isVisual bool, base64Image string) {"""

new_state = """type SearchState struct {
	Query       string
	Aspect      string
	Count       int
	IsVisual    bool
	Base64Image string
	Bookmark    string
}

func SetLastSearch(chatID, query, aspect string, count int, isVisual bool, base64Image string, bookmark string) {"""
content = content.replace(target_state, new_state)

target_set = """	lastSearches[chatID] = SearchState{
		Query:       query,
		Aspect:      aspect,
		Count:       count,
		IsVisual:    isVisual,
		Base64Image: base64Image,
	}"""
new_set = """	lastSearches[chatID] = SearchState{
		Query:       query,
		Aspect:      aspect,
		Count:       count,
		IsVisual:    isVisual,
		Base64Image: base64Image,
		Bookmark:    bookmark,
	}"""
content = content.replace(target_set, new_set)

# 2. Update SearchPinterest to support loop and bookmark
old_search = """func SearchPinterest(query string, aspect string, count int) []PinResult {
	query = url.QueryEscape(query)
	pageSize := count + 10
	if pageSize > 100 {
		pageSize = 100
	}
	searchUrl := fmt.Sprintf("https://api.pinterest.com/v3/search/pins/?rs=typed&pinrep_img_width=474x&query=%s&page_size=%d", query, pageSize)
	req, _ := http.NewRequest("GET", searchUrl, nil)
	setPinterestHeaders(req)

	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return nil
	}
	defer resp.Body.Close()

	bodyBytes, _ := ioutil.ReadAll(resp.Body)
	data := extractDataFromJSON(bodyBytes)

	return parsePinterestData(data, aspect)
}"""

new_search = """func SearchPinterest(query string, aspect string, count int, startBookmark string) ([]PinResult, string) {
	query = url.QueryEscape(query)
	var allPins []PinResult
	currentBookmark := startBookmark

	for len(allPins) < count {
		pageSize := 100
		searchUrl := fmt.Sprintf("https://api.pinterest.com/v3/search/pins/?rs=typed&pinrep_img_width=474x&query=%s&page_size=%d", query, pageSize)
		if currentBookmark != "" {
			searchUrl += "&bookmark=" + currentBookmark
		}

		req, _ := http.NewRequest("GET", searchUrl, nil)
		setPinterestHeaders(req)

		client := &http.Client{Timeout: 10 * time.Second}
		resp, err := client.Do(req)
		if err != nil {
			break
		}

		bodyBytes, _ := ioutil.ReadAll(resp.Body)
		resp.Body.Close()

		var respJson map[string]interface{}
		json.Unmarshal(bodyBytes, &respJson)

		var data []interface{}
		if d, ok := respJson["data"].([]interface{}); ok {
			data = d
		} else if d, ok := respJson["data"].(map[string]interface{}); ok {
			if res, ok := d["results"].([]interface{}); ok {
				data = res
			}
		}

		newPins := parsePinterestData(data, aspect)
		allPins = append(allPins, newPins...)

		if bm, ok := respJson["bookmark"].(string); ok && bm != "" && bm != currentBookmark {
			currentBookmark = bm
		} else {
			break
		}

		if len(newPins) == 0 {
			break
		}
	}

	return allPins, currentBookmark
}"""

content = content.replace(old_search, new_search)

with open("internal/pinterest/pinterest.go", "w") as f:
    f.write(content)
